text/0162-refactoring-support.md

```diff
@@ -0,0 +1,702 @@
+relevant stage of your pipeline.
+
+However, this only works if all versions are deployed. In many pipeline
+implementations, some versions may be skipped for various reasons. In such
+cases, the mapping files may become outdated, assuming a source state that is no
+longer accurate. Here are some examples:
+
+**Scenario 1**: Some version of your CDK application creates a resource with a
+certain logical ID that a later refactor references. But that version was
+skipped, and consequently the resource was never deployed. The refactor, then,
+references something that doesn't exist, which results in a CLI error, blocking
+the pipeline:
+
+```
+Action:   [Create "A"]  [Rename "A" to "B"]
+               │             │      
+Time:      ────┼─────────────┼────────────>
+               │             │      
+Deployed?      No            Yes
+                      ("A" doesn't exist)
+```
+
+**Scenario 2**: Due to skipped deployments, the refactor may end up referring to
+logical IDs that refer to different resources than intended. Suppose that, at
+some point, you had a resource called "A". In subsequent versions, "A" was
+deleted, and then replaced with a different resource that happened to also be
+named "A". If a mapping file references "A", the intention would be the most
+recent resource. But the resource that is found is the old one. In this case,
+the refactor will happen, but with a different result than intended:
+
+```
+Action:   [Create "A"]  [Delete "A"]  [Create diff. "A"]  [Rename "A" to "B"]
+               │             │          │                     │
+Time:      ────┼─────────────┼──────────┼─────────────────────┼─────────────>
+               │             │          │                     │
+Deployed?     Yes            No         No                   Yes
+                                                 (Applies to wrong resource)
+```
+
+If this is the case in your CI/CD pipeline, we don't recommend you use this
+feature.
+
+#### Ambiguity
+
+You may find yourself in the very unlikely situation in which two or more
+resources have the same type and properties, and they are moved or renamed at
+the same time. Since there are at least two valid ways to map the old IDs to the
+new ones, the CLI can't guarantee that it will make the right choice. For
+instance, suppose you have two identical S3 buckets, with logical IDs
+`FinancialData` and `CustomerData`, and you want to rename them to
+`FinancialReports` and `CustomerInfo`, respectively. But the CLI could end up
+doing the opposite, so that, after the refactor, the bucket that contains
+financial reports would be called `CustomerData` in CloudFormation, and vice
+versa.
+
+Even though all the resources involved in this scenario would remain unchanged,
+we have decided to err on the side of caution and not perform the refactor. If,
+as the result of computing a refactor, the CLI detects such a case of ambiguity,
+it will fail with an error message explaining the situation.
```

Comment:
I am unclear if we could output a potential mapping file with --record-resource-mapping, edit it and then use that to do the correct mapping, of whether that will also stop before creating the output file? (maybe the output file should have some warnings about the ambiguity/needs human review).

I assume the ambiguity can be resolved by supplying the mapping file (however obtained), please make that clear.